File: snippets/go/topological_sort.go
package go_snippets

type TopologicalSort struct {
	edges  [][]int
	sorted []int
	length []int // 入次数 0 のノードからそのノードまでの長さ
}

func (t *TopologicalSort) init(n int) {
	t.edges = make([][]int, n)
	t.length = make([]int, n)
}

func (t *TopologicalSort) add(from, to int) {
	t.edges[from] = append(t.edges[from], to)
}

// O(V + E)
func (t *TopologicalSort) sort() (res []int, isLoop bool) {
	inCount := make([]int, len(t.edges))

	for _, edgesForNode := range t.edges {
		for _, to := range edgesForNode {
			inCount[to]++
		}
	}

	// inCount が 0 のものの集合
	s := []int{}

	for node, in := range inCount {
		if in == 0 {
			s = append(s, node)
		}
	}

	for i := range t.length {
		t.length[i] = 1
	}

	for len(s) > 0 {
		popped := s[0]
		res = append(res, popped)
		s = s[1:len(s)]

		for _, to := range t.edges[popped] {
			if t.length[popped]+1 > t.length[to] {
				t.length[to] = t.length[popped] + 1
			}

			inCount[to]--

			if inCount[to] == 0 {
				s = append(s, to)
			}
		}
	}

	if len(res) != len(t.edges) {
		isLoop = true
		return
	}

	t.sorted = res

	return
}

File: snippets/cpp/topological_sort.cpp
#include "algorithm"
#include "iostream"
#include "set"
#include "vector"
#define rep(i, to) for (ll i = 0; i < (to); ++i)
using namespace std;
template <typename T>
using V = vector<T>;
using VL = V<ll>;
using VVL = V<VL>;
typedef long long ll;
template <typename T>
inline bool chmax(T& a, T b);

struct TopologicalSort {
  VVL edges;
  VL sorted, length;

  TopologicalSort(ll n) : edges(VVL(n, VL(0))), length(VL(n, 1)) {}

  void add(ll from, ll to) { edges[from].push_back(to); }

  // 戻り値: isLoop
  bool sort() {
    ll n = edges.size();
    VL in_count(n);

    rep(i, n) {
      for (auto to : edges[n]) in_count[to]++;
    }

    set<ll> in_count_zeros;

    rep(i, n) {
      if (in_count[i] == 0) in_count_zeros.insert(i);
    }

    VL res(0);

    while (in_count_zeros.size() > 0) {
      ll popped = *in_count_zeros.begin();

      res.push_back(popped);
      in_count_zeros.erase(popped);

      for (auto to : edges[popped]) {
        chmax(length[to], length[popped] + 1);

        in_count[to]--;

        if (in_count[to] == 0) in_count_zeros.insert(to);
      }
    }

    sorted = res;

    return res.size() != edges.size();
  }
};

template <typename T>
inline bool chmax(T& a, T b) {
  if (a < b) {
    a = b;
    return 1;
  }
  return 0;
}
